{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPukrDeI74dSp4hN9/XIkFw",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/himabindu-github/Hackbio-Oct25/blob/main/stage1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7eEk8n47vThp"
      },
      "outputs": [],
      "source": [
        "# Define the codon table\n",
        "dna_codon_table = {\n",
        "    \"TTT\": \"F\", \"TTC\": \"F\", \"TTA\": \"L\", \"TTG\": \"L\",\n",
        "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\", \"ATG\": \"M\",\n",
        "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
        "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"TAT\": \"Y\", \"TAC\": \"Y\", \"TAA\": \"*\", \"TAG\": \"*\",\n",
        "    \"CAT\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"AAT\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"GAT\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"TGT\": \"C\", \"TGC\": \"C\", \"TGA\": \"*\", \"TGG\": \"W\",\n",
        "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
        "    \"AGT\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"\n",
        "}\n",
        "\n",
        "\n",
        "# Define translation function\n",
        "def translate_dna(dna_seq):\n",
        "    dna_seq = dna_seq.upper()   # converts the sequence to uppercase to ensure consistency\n",
        "    protein = []                # start an empty list to store translated aminoacids\n",
        "    codons = [dna_seq[i:i+3] for i in range(0, len(dna_seq), 3)]  # Splits the DNA sequence into codons (3-base chunks) by taking every 3 nucleotides as one element in a list.\n",
        "\n",
        "    for codon in codons:   # loops through each codon\n",
        "        if len(codon) < 3:  # skip incomplete codons at the end\n",
        "            continue\n",
        "        elif codon in ['TAA', 'TGA', 'TAG']:  # stop translation if a stop codon is found\n",
        "            break\n",
        "        elif codon in dna_codon_table:\n",
        "            protein.append(dna_codon_table[codon]) # if codon exists in the dna_codon_table, add its aminoacid to the 'protein' list\n",
        "        else:\n",
        "            protein.append(\"X\")  # if unknown codon(characters other rhan A,T,G,C...like N), append 'X' to indicate an unknown amino acid\n",
        "\n",
        "    return \"\".join(protein) # join all amino acids in the list into a single protein sequence string\n",
        "\n",
        "\n",
        "print(\"Protein sequence :\", translate_dna(\"AGGTTGCCTTTA\")) # prints the translated protein for the given DNA sequence\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# In bioinformatics: Hamming distance measures the number of differing positions\n",
        "# between two DNA, RNA, or protein sequences of equal length.\n",
        "# useful for comparing genetic variants or detecting mutations.\n",
        "\n",
        "# defining a function to calculate Hamming distance between two sequences\n",
        "def hamming_distance(slack_name, x_name):\n",
        "    if len(slack_name) != len(x_name): # Check if both strings are the same length\n",
        "        raise ValueError(\"String mismatch\")  # # Raise an error if lengths are different\n",
        "    distance = sum(ch1 != ch2 for ch1, ch2 in zip(slack_name, x_name))\n",
        "    # zip: pairs up corresponding characters from both strings and compare them\n",
        "    # for each position, ch1 != ch2 returns True (1) if characters differ, False (0) if same\n",
        "    # adding up these True values gives the total count of differing positions\n",
        "    print(f\"Hamming distance = {distance}\")\n",
        "    # prints the result showing how many characters are different\n",
        "\n",
        "\n",
        "# output with 2 strings\n",
        "hamming_distance(\"himabindu\", \"raghurams\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u-woyhD8yF20",
        "outputId": "47beeaf2-4fe0-4765-f0cc-92a409ab8706"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Hamming distance = 9\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "eqsiYunnz0zg"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
